File: src/main/java/eotree/EOBndExpr.java
package eotree;

import java.util.Arrays;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class EOBndExpr extends EOBnd {
    public EOBndName bndName;

    public EOBndExpr(EOExpr expr, EOBndName bndName) {
        super(expr);
        this.bndName = bndName;
    }

    @Override
    public String generateEO(int indent) {
        var lines = expr.generateEO(indent).split("\n");
        // Append attribute name to the first line
        return Stream.concat(
                Arrays.stream(lines)
                        .findFirst()
                        .map(line -> line + " > " + bndName.generateEO(0))
                        .stream(),
                Arrays.stream(lines)
                        .skip(1)
        ).collect(Collectors.joining("\n"));
    }
}

File: src/main/java/eotree/EOBndExpr.kt
package eotree

import arrow.core.prependTo
import arrow.core.tail

class EOBndExpr(expr: EOExpr, var bndName: String) : EOBnd(expr) {
    override fun generateEO(indent: Int): String {
        val lines = expr.generateEO(indent).split("\n")

        val commentLines: List<String> = lines.takeWhile { it.trimStart().startsWith("#") }
        val nonCommentLines: List<String> = lines.dropWhile { it.trimStart().startsWith("#") }

        // Append attribute name to the first non-comment line
        return (
            commentLines +
                nonCommentLines
                    .first { !it.trimStart().startsWith("#") }
                    .let { line -> "$line > $bndName" }
                    .prependTo(nonCommentLines.tail())
            ).joinToString("\n")
    }

    override fun toString(): String = "[expr] > $bndName"
}
